### ISSUE TITLE
NSFW! gaybeeg.info, netdna-storage.com

### ISSUE THREAD

**Issue by phineasmax** [jspenguin2017/AdBlockProtector#164]
Thanks for your update for gaybeeg.info. It works perfectly.
But netdna-storage.com still detected adblock
Example: http://netdna-storage.com/f/RLESMQXo/FX-FreshCollegeMeat.mp4.html
![1](https://cloud.githubusercontent.com/assets/21326718/25381567/7dedf1de-29de-11e7-9a7a-045d700563dc.png)

**Comment by uBlock-user** [comment]
same as `gaybeeg.info` but with a race condition.

**Comment by jspenguin2017** [comment]
@uBlock-user Hum, I don't think it's race condition... I found the detection script, they got smarter this time... `_blocked` is now in a closure... 
```JavaScript
    (function() {
    	var _blocked = 0;

    	function injectScript(src) {
			var script = document.createElement('script');
			script.async = true;
			script.src = src;
			script.addEventListener('error', function () {
				_blocked++;
			});
			document.head.appendChild(script);
		}

		[
			'//www.adnetworkperformance.com/a/display.php?r=1151825'
		].map(function(src) {
			injectScript(src);
		});

		$(function(){
			if(_blocked > 0){
				$('.warning').show();
				$('.plan-footer-item').removeClass('btn--grid').removeAttr('data-href');
			} else {
				$('.plan-footer-item').each(function(i, el){
					$(el).attr('href', $(el).data('href')).removeAttr('data-href');
				});
			}
		});
    })();
```

**Comment by uBlock-user** [comment]
> _blocked is now in a closure...

Killing `injectScript` should work too, but it doesn't, can you justify that ?

**Comment by jspenguin2017** [comment]
@uBlock-user As I said before, it's in a closure.

**Comment by uBlock-user** [comment]
You said `_blocked` is in a closure, I'm talking about `injectScript`, if I use `a.noAccess` on it, it should break the script.

**Comment by jspenguin2017** [comment]
@uBlock-user Well, `_blocked` and `injectScript` are both in the closure.

**Comment by jspenguin2017** [comment]
Hum, I think patching `Element.prototype.addEventListener` would work... The Core has some design flaw, that needs to be fixed before we can proceed...

**Comment by uBlock-user** [comment]
FYI jQuery 3.2.1 was recently released, shou;dn't you update the version to the latest ?

**Comment by uBlock-user** [comment]
Link - http://blog.jquery.com/

**Comment by jspenguin2017** [comment]
@uBlock-user 
OK, thanks for letting me know. I don't think there is any rush in updating it.

**Comment by jspenguin2017** [comment]
This line would break their detection, along with everything else. So no, it's definitely not race condition. We need to patch Core first, this can take some time. 
```JavaScript
a.win.Element.prototype.addEventListener = null;
```

**Comment by phineasmax** [comment]
I really appreciate your works. That site also have some kind of IP tracking along with adblock detector. Hope you guys can make it work.

**Comment by uBlock-user** [comment]
I suggested that because you mentioned The Core has some design flaw, so updating jQuery should patch the flaw too.

**Comment by jspenguin2017** [comment]
@uBlock-user 
Well, I guess... But the flaw we need to patch doesn't have anything to do with jQuery. 
The flaw is the way element properties are processed, we need to loop that so we can handle an arbitrarily deep property.

**Comment by jspenguin2017** [comment]
OK, I hacked a solution together, this should do the job until we can properly fix it. It is added to Script
```JavaScript
if (a.domCmp(["netdna-storage.com"])) {
    //Temporary fix
    //Need to patch Core so this works: 
    //a.filter("Element.prototype.addEventListener", "script");
    const _createElement = a.doc.createElement;
    a.doc.createElement = function (what) {
        if (what === "script") {
            let elem = _createElement.apply(a.doc, arguments);
            elem.addEventListener("error", function (e) {
                e.stopPropagation();
            });
            return elem;
        } else {
            return _createElement.apply(a.doc, arguments);
        }
    }
}
```

**Comment by uBlock-user** [comment]
Unfortunately, no change on my end.

**Comment by jspenguin2017** [comment]
It's working for me... 
![image](https://cloud.githubusercontent.com/assets/7283682/25394062/7075fee6-299a-11e7-8f6e-6041e66c0114.png)

**Comment by uBlock-user** [comment]
![](https://i.gyazo.com/f2aa688b0ed2a1487a2ab8dcac801749.png)

**Comment by jspenguin2017** [comment]
What's your Script version? Can you dump debug log?

**Comment by uBlock-user** [comment]
v6.187

![](https://i.gyazo.com/cb43b0e07d79a5f75c62766e28deb369.png)

**Comment by ghajini** [comment]
race condition....1st time it worked after updated to 6.187....again checked,now it DOESN'T work

**Comment by jspenguin2017** [comment]
Hum... How about nooping their request?

**Comment by jspenguin2017** [comment]
This works for me: 
`||adnetworkperformance.com^$script,redirect=noopjs,domain=netdna-storage.com`

**Comment by uBlock-user** [comment]
nooping the ad -server ? I won't do it as I don't want to entertain any requests from an ad-server. Besides, they can just change the ad-server like debridinet.com did after gorhill nooped the adframe from `amazon-adserver`.

**Comment by jspenguin2017** [comment]
@uBlock-user So what's the alternative?

**Comment by uBlock-user** [comment]
I don't know, are you out of all options ? Can't you patch it ?

**Comment by uBlock-user** [comment]
Blocking inline-script kills the anti adblock notice and the download link too.

**Comment by uBlock-user** [comment]
This also kills the anti adblock notice - `a.readOnly("$", function () {});`

**Comment by jspenguin2017** [comment]
If I'm not out of options, I wouldn't choose to noop connections. 
Some Adfly variables are assigned in-line and blow up jQuery usually don't go well. Unless you can make it actually work, I'd say we are out of options.

**Comment by uBlock-user** [comment]
If you just want a patch, then go ahead and noop it and add it to the List, as the worst part is the download link takes you to other shit url shorteners and your download will never begin as it will keep rotating and I have no reason to visit this website ever.

**Comment by jspenguin2017** [comment]
Well, that's how the website is designed. I would never go visit it beside handling issues like this one, I agree on this. 
I chose to noop it because there is no other simple way. If anyone found a better working solution, please let me know.

**Comment by lain566** [comment]
It's working for me on first page only

**Comment by jspenguin2017** [comment]
I think those guys are refactoring their code, because yesterday this website is working without any rule, this morning it is working with some nooping, and now it broke again and it has some extra text offering lower quality video for small data plan...

**Comment by jspenguin2017** [comment]
Hum, it requests other files on the other pages... I nooped them too. 
The files that are requested are actually executed, so I guess they can't just swap them around. I hope nooping can last reasonable amount of time.

**Comment by lain566** [comment]
the second page (step 2) not working 

![2017-04-25_22-55-39](https://cloud.githubusercontent.com/assets/27441315/25407568/0db698b2-2a0b-11e7-9f25-4d5deccec0e4.gif)

**Comment by jspenguin2017** [comment]
Did you update your filter? I can't reproduce that after the update. 
![image](https://cloud.githubusercontent.com/assets/7283682/25409959/7ceee12e-29d1-11e7-83a6-02617083ff7a.png)

**Comment by lain566** [comment]
ohhh nice
its works!

![watch_-_download_fx-freshcollegemeat mp4_-_google__2017-04-26_00-13-45](https://cloud.githubusercontent.com/assets/27441315/25410146/5fd84320-2a15-11e7-9a70-8358774665a7.png)

**Comment by phineasmax** [comment]
![untitled](https://cloud.githubusercontent.com/assets/21326718/25415586/d86c9ab4-2a62-11e7-9ec2-03e3ed9968ef.png)
That's right. It works.

**Comment by lain566** [comment]
@jspenguin2017 Sorry for the screenshot, i did not know it was xxx xxx. lol
Well, I found this,  maybe this is better solution.
https://github.com/JonnyShuali/netdna-storage-click-bypass

**Comment by jspenguin2017** [comment]
@lain566 Hum, interesting... Only 4 lines of code: 
```JavaScript
var downBtn = document.getElementsByClassName('plan-footer-item');
if (downBtn.length === 1) {
  downBtn[0].click();
}
```
Did you test it? Does it actually work?

**Comment by jspenguin2017** [comment]
@lain566 It seems that that code simply click on the "Download" and "Continue" buttons for you, which isn't really useful since it doesn't hop over those "url shrinker". We have a solution that goes right through Adfly, but that's about it, it only works for Adfly.

**Comment by lain566** [comment]
@jspenguin2017 Only works with AdsBypasser,  is not updated since 1 year ago

**Comment by jspenguin2017** [comment]
@lain566 Does AdsBypasser skip through all the layers and directly put you on download page? Or you still have to go through all 4 steps?

**Comment by jspenguin2017** [comment]
Ah, I see... All it does is clicking the button for you, which I don't think is very useful... We do have part of AdsBypasser integrated, but it doesn't cover all the intermediate steps. I'll be too long and complex to implement more, most of them are simply wait 3 or 5 seconds. 
Do you think it'll be helpful to make the click automatic?

**Comment by lain566** [comment]
I think this would be helpful to adsbypasser

**Comment by jspenguin2017** [comment]
@lain566 In that case, it would be better to submit a request to AdsBypasser.

**Comment by lain566** [comment]
they work only on countdown pages

**Comment by jspenguin2017** [comment]
OK, in that case, I'll implement it.

**Comment by lain566** [comment]
Are you going to add bypass for shorte.st and ouo.io??

**Comment by jspenguin2017** [comment]
@lain566 Probably I'll just `timewarp` the timer, I'll check it out.

**Comment by lain566** [comment]
http://clkmein.com/qICV9i

**Comment by jspenguin2017** [comment]
@lain566 Awesome! Do you have some test links for `ouo.io`?

**Comment by lain566** [comment]
ouo domains:
ouo.io/qR1Fn
ouo.press/qR1Fn

**Comment by ghajini** [comment]
No need of bypassing countdown timers since it doesn't trigger adblock......don't make things complex @jspenguin2017

**Comment by jspenguin2017** [comment]
@ghajini We already have `a.timewarp()` is the Core, it's not going to be complex... I hope...

**Comment by uBlock-user** [comment]
Didn't nooping  the adserver patch the website or are we still looking for solutions ?

**Comment by uBlock-user** [comment]
Also countdown timers are put in the lieu of making things worse for adblock users, so it should be patched, it's nowhere near complex.

**Comment by jspenguin2017** [comment]
@uBlock-user It patched the website, but we are trying to make the process faster by skipping countdowns and clicking on the button programmatically.

**Comment by phineasmax** [comment]
Okay. The issue of gaybeeg has been moved here.
I tried to disable all addons, just keep uBlock and Tampermonkey.
Still detected adblocker and redirect me to the adblock site (http://gaybeeg.info/adblock-instructions/)
I believe that the site (gaybeeg.info) when they detected my adblocker, they listed my IP to their blacklist and they redirect all the download link to the instructions link. I tried to use VPN, it worked for the first time and when I refresh the page, it detected and redirect again. Tried to change IP for many times and it still like that: worked for the first time.

**Comment by jspenguin2017** [comment]
@phineasmax Hum, I see. I guess it's the cookie... We'll fix it.

**Comment by phineasmax** [comment]
@jspenguin2017 Thanks.
I have a demo here. Sorry about the quality. (have to resize)
![untitled-3](https://cloud.githubusercontent.com/assets/21326718/25561614/3d454aaa-2d9a-11e7-9697-7d33d7140b0d.gif)

**Comment by jspenguin2017** [comment]
@phineasmax OK, I see the problem, `gaybeeg.info` and `netdna-storage.com` are pretty much the same website. The detection script is really the same. They seem to be changing their detection script daily for the last week. 
This is the current script, looks like we need to send a XML request. 
<details>

```JavaScript
var uid = '61264';
var wid = '111331';

// PopAds.net config
var _pop = _pop || [];
	_pop.push(['siteId', 570806]);
	_pop.push(['minBid', 0]);
	_pop.push(['popundersPerIP', 0]);
	_pop.push(['delayBetween', 0]);
	_pop.push(['default', false]);
	_pop.push(['defaultPerDay', 0]);
	_pop.push(['topmostLayer', false]);


function tellnetdna(status) {
	var request = new XMLHttpRequest();
	request.open('POST', 'http://netdna-storage.com/api');
	request.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded; charset=UTF-8');
	request.send(encodeURI('status=' + status));
}



(function($){


	var _9614cb7 = 0;


	function injectScript(src) {
		var script = document.createElement('script');
		script.async = true;
		script.src = src;
		script.addEventListener('error', function () {
			_9614cb7++;
		});
		document.head.appendChild(script);
	}


	[
		'//c1.popads.net/pop.js',
		'//do67etikr7pwz.cloudfront.net/?iteod=604791',
		'//cdn.popcash.net/pop.js',
		'//syndication.exoclick.com/splash.php?idzone=1551172&type=3'
	].map(function(src) {
		injectScript(src);
	});


	$(function(){

		if(_9614cb7 > 0){

			$('.download a.button').each(function(i, el){
				$(el).removeAttr('data-href').attr('href', 'http://gaybeeg.info/adblock-instructions/');
			});

	 		$('body').append('<div class=\"msc-confirm\"><div class=\"msc-overlay\"></div><div class=\"msc-content msc-confirm--animate\"><h3 class=\"msc-title\">AdBloker Detected</h3><div class=\"msc-body\"><p class=\"msc-sub\">Looks like you might have an AdBlocker on.</p><p class=\"msc-sub\">You should either <a class=\"whitelist\" href=\"http://gaybeeg.info/adblock-instructions/\" target=\"_blank\">whitelist</a> GayBeeg.info OR become Premium, to enjoy an Ad-free experience.</p></div><div class=\"msc-action\"><a class="msc-ok" href="http://gaybeeg.info/login/?redirect_to=' + encodeURIComponent(window.location.href) + '">Login</a></div></div></div>');

	 		tellnetdna(true);
		} else {

			$('.download a.button').each(function(i, el){
				$(el).removeClass('locked').attr('href', $(el).data('href')).removeAttr('data-href');
			});

			tellnetdna(false);
		}
	});
})(jQuery);
```

</details>
Also, which software do you use to create those GIFs?

**Comment by phineasmax** [comment]
@jspenguin2017 I used xbox app to record the screen and Adobe Photoshop to create the gif.
Hope you guys can make it works.

**Comment by jspenguin2017** [comment]
@phineasmax I mean, it's 5 minutes of work to get it to work, but @uBlock-user was right, we need to make patches to be hard enough to bypass that it can last months, not days. 
It's a bit late here, I have a plan, I'll do it tomorrow.

**Comment by uBlock-user** [comment]
Usage of  jQuery is really an issue, isn't it ? There's really not much we can do about it as I have noticed.

**Comment by phineasmax** [comment]
@jspenguin2017 It's okay. Stable is the most important here. You don't have to rush.

**Comment by jspenguin2017** [comment]
@uBlock-user Obviously jQuery isn't the problem, don't worry, I'll update it, but because it's a bit more complex than just copy pasting the new code, it's not a priority. 
And there is a lot we can do, we can crash the script when it's assigning `_pop` and call `tellnetdna` ourselves. Because it is function declaration, it runs first, even before assignment of `uid` is done. The problem? They can move `tellnetdna` into the closure and that is 5 seconds of work with no chance to break other things. 

 @lain566 Timer skipping is separated to other issues, this one will be dedicated to `gaybeeg.info` and `netdna-storage.com`

**Comment by uBlock-user** [comment]
@jspenguin2017 This is why I suggested blocking the `inline-script`, so we wouldn't have to deal with webmaster's desperate and pathetic attempts at self-sabotage and given your existing solution in the script, the links worked just fine and the website wasn't broken either.

**Comment by jspenguin2017** [comment]
@uBlock-user Nope, they have server side code, you got to fire that XHR. And I'm afraid that we would need some serious monkey patching to get Adfly working.

**Comment by jspenguin2017** [comment]
@uBlock-user Also, blocking all in-line scripts does break the site a little bit, the archive section doesn't work anymore. And from inspecting their code, I think some emoji will stop working as well.

**Comment by uBlock-user** [comment]
>  you got to fire that XHR

What do you mean ? I'm talking about `gaybeeg.info` 

Secondly the only thing there to do is open the website, click the blog post, checkout the screenshots and if the user likes the porn, then he can click on the download which leads to the `netdna-storage` domain.

**Comment by uBlock-user** [comment]
Emoji where ? In the comment section ? I haven't posted any comments to check that out, did you ?

**Comment by uBlock-user** [comment]
The problem is the over-zealous webmaster who will undo your solution within hours, and you will be back to play his `whack-a-mole` game.

**Comment by jspenguin2017** [comment]
@uBlock-user You can't assume what other people do, we can't break archive and other stuff assuming it is optional. Trust me, I know what I'm doing. I'll fix it.

**Comment by uBlock-user** [comment]
I never said it's optional. I meant choosing between breaking a dropdown button of the archive section or playing his game everyday. I'm sure you can fix it, you already did twice, the question is how long will you keep up.

**Comment by jspenguin2017** [comment]
@uBlock-user It'll keep up until they change the way their API callback works. Go try it out, with current solution, it only works for the first visit, they implemented something on their server side.

**Comment by uBlock-user** [comment]
> Go in the page, wait for it to load, refresh, and you'll see all the links are broken

Nope, as long as I have Script enabled, links work fine, no breakage on my end. If I disable the Script, then only the links remain broken.

**Comment by jspenguin2017** [comment]
@uBlock-user If you click the link, where does it send you?

**Comment by jspenguin2017** [comment]
@uBlock-user I want to fix the archive.

**Comment by uBlock-user** [comment]
I don't understand. Are you using the inline-script filter ? Otherwise the archive doesn't break.

**Comment by jspenguin2017** [comment]
@uBlock-user Sorry if that was unclear, so here is the situation: 

* Use in-line script blocking: Archive breaks
* Not using in-line script blocking: Refresh and links break

**Comment by uBlock-user** [comment]
> Use in-line script blocking: Archive breaks

Yes, I agree.

> Not using in-line script blocking: Refresh and links break

Not exactly, links do break on my end, but as I mentioned before, that happened in alternate ways, they break the first time, once I click refresh and the links are fixed on my end.

**Comment by jspenguin2017** [comment]
@uBlock-user That doesn't agree with my tests. Maybe the API is blocked somewhere on your end. Anyway, I fixed it. Block in-line scripts and use hash to run selected code. 
```JavaScript
if (a.domCmp(["gaybeeg.info"])) {
    a.observe("insert", function (node) {
        if (node.innerHTML && node.innerHTML.includes("AdBloker Detected")) {
            node.remove();
        }
    });
    a.ready(function () {
        a.$("script").each(function (i, elem) {
            if (!elem) {
                return;
            }
            if (a.hash(elem.innerHTML) === 861935216) {
                a.win.eval(elem.innerHTML);
                return;
            }
            if (elem.innerHTML.includes("/*  Collapse Functions, version 2.0")) {
                const temp = elem.innerHTML.split("/*  Collapse Functions, version 2.0");
                if (temp.length === 2 && a.hash(temp[1]) === -1728731483) {
                    a.win.eval(elem.innerHTML);
                    return;
                }
            }
            if (a.config.debugMode) {
                a.out.warn("This inline script is not executed: ")
                a.out.warn(elem.innerHTML);
                a.out.warn("Hash: " + a.hash(elem.innerHTML).toString());
            }
        })
        a.$(".download a.button").each(function (i, el) {
            a.$(el).removeClass("locked").attr("href", a.$(el).data("href"))
                .removeAttr("data-href");
        });
    });
}
```

**Comment by uBlock-user** [comment]
With that code, do I still need to block the inline-script ?

**Comment by jspenguin2017** [comment]
@uBlock-user Yes.

**Comment by jspenguin2017** [comment]
I just realized that my hash function is pretty flawed... I need to switch to CRC 32. 
https://github.com/SheetJS/js-crc32/blob/master/crc32.js

**Comment by uBlock-user** [comment]
SHA 1 is bad, collisions were found recently by Google, so avoid it and use SHA 2 or SHA 3.

**Comment by jspenguin2017** [comment]
@uBlock-user 
I wanted to use CRC 32 or SHA 1 because they are fast. For our purposes, collision attack could be possible on CRC 32, but not SHA 1. I don't think any reasonable human will burn 110 GPU years just to crack our checksum. 
But unfortunately I couldn't find any good optimized SHA 1 function for JavaScript. Which removes the only quality of SHA 1, so I ended up choosing SHA 256, which is a variant of SHA 2.

**Comment by phineasmax** [comment]
Hi. I just visit the site (gaybeeg.info) and the download buttons is grey, instead of green.

**Comment by uBlock-user** [comment]
@phineasmax Did you subscribe to the List ? Works fine here.

**Comment by phineasmax** [comment]
@uBlock-user I tried to install chrome and the script and the list as well. It works on Chrome.
But not firefox. I even tried to Uninstall all and reinstall firefox, the script and the list. But it didn't work. Keep redirecting me to the instructions site.
![screenshot 35](https://cloud.githubusercontent.com/assets/21326718/25781236/5c00a49a-3361-11e7-97a6-16c1301bb4c5.png)
(on the left: Chrome. You can see that's netdna....)
![screenshot 36](https://cloud.githubusercontent.com/assets/21326718/25781229/4d495a32-3361-11e7-9a9d-ecb2d0904d59.png)
(On the right: Firefox, redirect to the instructions site.)
One thing weird is that Chrome didn't block the popup, but firefox did.
and there is another different:
![screenshot 37](https://cloud.githubusercontent.com/assets/21326718/25781271/d2bb8df2-3361-11e7-92fe-a3555b8a023a.png)

**Comment by jspenguin2017** [comment]
@phineasmax You are right, it doesn't work on FireFox, the console isn't really working neither, I think it's a bug in FireFox.

**Comment by jspenguin2017** [comment]
@phineasmax Looks like FireFox isn't respecting Content Security Policy, try this filter: 
`gaybeeg.info##script:inject(abort-on-property-read.js, tellnetdna)`

**Comment by phineasmax** [comment]
Well. Added that filter and still grey. I didn't use Incognito mode this time.
Which bugs need to be reported?

**Comment by jspenguin2017** [comment]
@phineasmax If you are using FireFox, the bug is that FireFox does not respect Content Security Policy, it only blocks 2 inline scripts instead of all of them, this **might** be a bug in uBO or Tampermonkey, but since it works on Chrome, I assume FireFox is bugged.

**Comment by jspenguin2017** [comment]
Since this works on our supported setups, this is considered fixed so I'll close this issue now.
